### ISSUE TITLE
A shorter equivalent way to declare `<T extends @Nullable Object>`

### ISSUE THREAD

**Issue by kevinb9n** [jspecify/jspecify#289]
### The situation

The status quo isn't a disaster, but it sure ain't great:

![image](https://user-images.githubusercontent.com/934551/190724799-ec41bd77-e37b-4c07-b73a-487abb4c20b5.png)

As much of a pain as it is to write all that out, it bothers me much more that users will *see* it in Javadoc, and there's no real way around it.

### Candidate solution

`interface Foo<@SomeNewAnnotation T>`

The annotation name should not necessarily allude to the idea of *being nullable*, but should convey the idea that it "accepts *either* nullness", i.e. "is *actually* parameterized along the nullness axis at all".

This is for type variables only; it would be impossible to put before a wildcard.

If there is any explicit bound (including `Object`), I'm guessing we'd declare this annotation unrecognized and have the user do it the old-fashioned way; could be debated.

The current syntax would continue to work, and I'd assume we don't care whether a tool recommends switching to the shorter style or not.

### Advantages

* Easier to write
* Easier to read (in source or, importantly, in javadoc)
* Lets users get by with the "basic model" (bits of nullness info ride along with types, and checkers Do Smart Things with that) for a while longer before they have to understand the "type system model" (parallel hierarchies and non-null<:nullable; or union types; see #274)
* With the right name, it's more clear why classes like in #78 would not use it
* For a happy CF user considering whether to make their library JSpecify-friendly, adding `@NullMarked` changes very little other than forcing their `<@NonNull A>`'s to change to `<A>`s and their `<B>`'s to change to (this syntax). This feature would make that change more of a lateral move, less of a "do I even want to do this?"
* Subjectively JSpecify may feel less like it hates its users

### Drawbacks

* (YAA) Yet another annotation
* The best name we can find might still not be that good
* Two Ways To Do The Same Thing
* Suggests that type parameters are special, when really we're only (elegantly) augmenting types themselves

### Now or later?

This is marked Post-1.0. Some arguments for moving it up to 1.0:

* Some bad first impressions (that we obviously don't care about usability) might already have done their damage
* API owners can't safely use it until tool support has percolated all the way out to their users
* Some APIs would be left doing it the long way forever anyway

### Alternatives

Could be obviated by #82 or #12.

Having `@Nullable` itself serve this purpose is probably a non-starter since some tools will continue recognizing their own `@Nullable` annotation there *with a different meaning* (#85, which I still feel fine about rejecting)

**Comment by msridhar** [comment]
I am open to doing something here.  Maybe `interface Foo<@PossiblyNullable T>`?  I think the tradeoffs are spelled out pretty clearly above.  I do think that assuming the rules don't end up being super complex, this a relatively minor ask for tool owners.

**Comment by cpovirk** [comment]
> * since we aren't proposing to let users write `interface Foo<@SomeNewAnnotation T extends Number>`

Oh, sorry, for the second time today, I was commenting on an earlier version of a post. It sounds like you're proposing to allow that after all? (As you say, it does sound like we'd be required to, since we bytecode doesn't distinguish between `interface Foo<@SomeNewAnnotation T>` and `interface Foo<@SomeNewAnnotation T extends Object>`.)

**Comment by kevinb9n** [comment]
Indeed, when analyzing dependent code we'd have to recognize `<@TheAnnotation T extends Object>`; then, it's equally clear what `<@TheAnnotation T extends NotObject>` is intended to mean, so it might as well recognize that too.

But when analyzing source, we would want tools to complain about using this together with any explicit bound. At best it's redundant (when the bound is `@Nullable`) and at worst the bound becomes an outright lie.

**Comment by kevinb9n** [comment]
> * The idea is more than "accepts either nullness," since we aren't proposing [edit: oops, I'm probably wrong, see below] to let users write `interface Foo<@SomeNewAnnotation T extends Number>` with behavior equivalent to `interface Foo<T extends @Nullable Number>`.
> * The idea is more than even "can accept any type argument, even nullable ones": In `interface Foo<T extends @Nullable Object, U extends T>`, `U` "can (sometimes) accept any type argument, even nullable ones," but we wouldn't want `@SomeNewAnnotation` there.

It sounds to me like when `@SomeNewAnnotation` is used, it *does* mean those things; it's just that when your goal is to "mean those things", `@SomeNewAnnotation` isn't *always* the way to do it. Fair? I don't think that bothers me tooo much.

**Comment by cpovirk** [comment]
Yes, that's very fair: `@SomeNewAnnotation` means those things; the absence of `@SomeNewAnnotation` doesn't mean the absence of those things.

I still find the drawbacks more compelling than the advantages overall (+1 to the comment that you lay them out clearly here), but I retract the additional drawback that I'd been describing above.

**Comment by scolsen** [comment]
It's too bad we can't overload `Nullable@`, imo this approach both makes a lot of intuitive sense and keeps our ontology of annotations smaller.

Another idea I think would be so-so is to move the annotation one level up and apply it at the type declaration level, but that wouldn't show up in a useful way in javadoc, right? something like: `@NullableTypeParams interface <T>`.

**Comment by kevinb9n** [comment]
* I feel that if `class Foo<@Nullable T>` is valid at all then #85 is the only reasonable meaning for it to have. If it had the meaning we're discussing here, it would be super mysterious why `T` isn't nullable and only the apparently-redundant `@Nullable T` is.
* Hmm, to me it feels too hacky to pretend this is a property of the class, even assuming it's extremely rare for a class's type params to be of mixed bounds.

**Comment by kevinb9n** [comment]
If there is any possibility of `class Foo<@SomeNewAnnotation T>`, perhaps a palatable name for it would be `class Foo<@Any T>`. I realize this doesn't mesh with how other languages use "any". But, maybe we can at least downplay "the name is so long and ugly" from our list of disadvantages of the approach.

**Comment by cpovirk** [comment]
I see how that name would be used only in Java sources, sidestepping concerns about the Kotlin `Any`. However, I happen to know from [bitter experience](https://github.com/protocolbuffers/protobuf/blob/242771ad159927cf731f30c1379aaef1913d7b1e/src/google/protobuf/compiler/java/message.cc#L1380) that there is _another_ commonly used `Any` class, [`com.google.protobuf.Any`](https://protobuf.dev/reference/java/api-docs/com/google/protobuf/Any). That probably makes the name a non-starter for us.

**Comment by kevinb9n** [comment]
I actually think the idea of someone maintaining a javadoc postprocessor is... not terrible, given all the room for improvement. But I would be surprised if someone signs up for it!

For JSpecify we should assume it is what it is, and take that into account.

**Comment by agentgt** [comment]
Have you guys used Eclipse null analysis lately?

![eclipse-nonnull](https://github.com/jspecify/jspecify/assets/567706/e0aba81f-69ba-43a1-b8d2-90e3ab73a979)

So I agree with @sviperll [points on user pain (reddit)](https://www.reddit.com/r/java/comments/13wxcti/design_document_on_nullability_and_value_types/jml4m5s/) but clearly Eclipse has no problem shoving `@NonNull/@Nullable` everywhere (actually it is worse in that it will actually attach the `@NonNull` annotation to the type when Eclipse compiles even thought it isn't actually annotated thus when you go check with an annotation processor you will see the annotation present! ... There is a bug filed somewhere for that).

Yes I am probably the last person still using Eclipse but I have gotten used to the above and luckily this only happens in the IDE. 

In great irony to the above Eclipse [does allow you to specify if generics are to be](https://javadoc.io/static/org.eclipse.jdt/org.eclipse.jdt.annotation/2.0.0/org/eclipse/jdt/annotation/DefaultLocation.html#TYPE_PARAMETER) interpreted as `@NonNull` or unspecified (unspecified is more or less `T extends @Nullable Object`). 

Let me tell you that feature is great for older applications (it helped me get our whole codebase nonnull) but relatively ... pardon my french confusing as fuck later when you try to remember what you have flipped on and off.

IntelliJ will do interesting thing as well. If you have it try to infer and auto annotate your code (I have never gotten that feature to work) it will just `@NotNull` on everything.

So for me I don't mind the extra step as a library author and or even the javadoc bloat (although I do like @bowbahdoe idea of using `?` for nullable... maybe Eclipse can use it instead for hover over as well as the possible future `!` for nonnull).

**Comment by cpovirk** [comment]
> actually it is worse in that it will actually attach the `@NonNull` annotation to the type when Eclipse compiles even thought it isn't actually annotated thus when you go check with an annotation processor you will see the annotation present! ... There is a bug filed somewhere for that

I've seen you before on conversations of https://bugs.eclipse.org/bugs/show_bug.cgi?id=572354, so probably that :)

**Comment by agentgt** [comment]
Luckily they (JDT) now has their issues on github: https://github.com/eclipse-jdt/eclipse.jdt.core/issues

The last time I used bugzilla besides Eclipse was when CVS was still an OK option. It was one of my biggest pain points of Eclipse. I wish even intellij moved over (albeit their bug system is much better than bugzilla).

I don't see that bug on github so I'll add it later.

**Comment by kevinb9n** [comment]
This belongs in the Post-1.0 bucket, rather than saying 'working decision: no'.  (You could say that we made a "working decision" to postpone it, but we aren't generally tracking things that way.)
